ibis/backends/duckdb/__init__.py

```diff
@@ -1609,7 +1609,10 @@ def _in_memory_table_exists(self, name: str) -> bool:
             return True
 
     def _register_in_memory_table(self, op: ops.InMemoryTable) -> None:
-        self.con.register(op.name, op.data.to_pyarrow(op.schema))
+        if hasattr(op.data, "to_pyarrow_dataset"):
+            self.con.register(op.name, op.data.to_pyarrow_dataset(op.schema))
+        else:
+            self.con.register(op.name, op.data.to_pyarrow(op.schema))
```

Comment:
Eventually we may want something like `to_native`, that returns the underlying object.

In the meantime, we can just try to call this and catch the attribute error.

```suggestion
        try:
            self.con.register(op.name, op.data.to_pyarrow_dataset(op.schema))
        except AttributeError:
            self.con.register(op.name, op.data.to_pyarrow(op.schema))
```